File: Solutions/src/Solution1.java
// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach

import java.util.HashMap;

class Solution1 {
    public int subarraySum(int[] nums, int k) {
        if(nums==null || nums.length==0){
            return 0;
        }

        HashMap<Integer,Integer> map = new HashMap<Integer,Integer>();
        int count=0;
        int rsum=0;
        map.put(0,1);
        for(int i=0;i<nums.length;i++){
            rsum=rsum+nums[i];
            if(map.containsKey(rsum-k)){
                count +=map.get(rsum-k);
            }
            if(!map.containsKey(rsum)){
                map.put(rsum,1);
            }
            else{
                map.put(rsum,map.get(rsum)+1);
            }
        }
        return count;
    }
}

File: solution1.py
# https://leetcode.com/problems/subarray-sum-equals-k
# // Time Complexity : o(n)
# // Space Complexity : o(n)
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this : got confused in case k==0, realized I need to increase count first then add the new element in hash
#

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        hm = dict()
        hm[0] = 1
        sum1 = []
        seed = 0
        count = 0

        for num in nums:
            temp = seed + num
            sum1.append(temp)
            seed = temp
        for i in range(len(sum1)):

            if sum1[i] - k in hm:
                count += hm[sum1[i] - k]

            if sum1[i] not in hm:
                hm[sum1[i]] = 1
            else:
                hm[sum1[i]] += 1
        return count
